### PR TITLE
Implement communication with Status JSON CGI

### PR DESCRIPTION
**Pull request by DanielSz50**

### CODE REVIEW

File: nagios/status.go [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+	q.SetNonEmpty("contactgroup", g.ContactGroup)
+	q.SetNonEmpty("checktimeperiodname", g.CheckTimeperiodName)
+	q.SetNonEmpty("hostnotificationtimeperiodname", g.HostNotificationTimeperiodName)
+	q.SetNonEmpty("checkcommandname", g.CheckCommandName)
+	q.SetNonEmpty("eventhandlername", g.EventHandlerName)
+	q.SetNonEmpty("contactname", g.ContactName)
+
+	q.SetNonEmpty("hosttimefield", g.HostTimeField)
+	q.URLQuery.Set("starttime", strconv.FormatInt(g.StartTime, 10))
+	q.URLQuery.Set("endtime", strconv.FormatInt(g.EndTime, 10))
```
Comment (amwolff) [inline nagios/status.go]: Status JSON CGI can have these fields empty.

File: nagios/status.go [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+
+type HostCount struct {
+	FormatVersion int           `json:"format_version"`
+	Result        Result        `json:"result"`
+	Data          HostCountData `json:"data"`
+}
+
+type HostListData struct {
+	Selectors map[string]json.RawMessage `json:"selectors"`
+	HostList  map[string]json.RawMessage `json:"hostlist"`
```
Comment (amwolff) [inline nagios/status.go:135]: Could it be unmarshaled into a `map[string]string` or something like that?

File: nagios/status.go [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+	q.SetNonEmpty("servicedescription", g.ServiceDescription)
+	q.SetNonEmpty("checktimeperiodname", g.CheckTimeperiodName)
+	q.SetNonEmpty("servicenotificationtimeperiodname", g.ServiceNotificationTimeperiodName)
+	q.SetNonEmpty("checkcommandname", g.CheckCommandName)
+	q.SetNonEmpty("eventhandlername", g.EventHandlerName)
+	q.SetNonEmpty("contactname", g.ContactName)
+
+	q.SetNonEmpty("servicetimefield", g.ServiceTimeField)
+	q.URLQuery.Set("starttime", strconv.FormatInt(g.StartTime, 10))
+	q.URLQuery.Set("endtime", strconv.FormatInt(g.EndTime, 10))
```
Comment (amwolff) [inline nagios/status.go]: Status JSON CGI can have these fields empty.

File: nagios/status.go [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+	q.SetNonEmpty("query", query)
+	q.SetNonEmpty("formatoptions", h.FormatOptions.String())
+	q.SetNonEmpty("dateformat", h.DateFormat)
+	q.SetNonEmpty("hostname", h.HostName)
+
+	return q
+}
+
+func (h HostRequest) Build() Query {
+	return h.build("host")
```
Comment (amwolff) [inline nagios/status.go]: You don't need the intermediate method here (`build` can be just `Build`).

File: nagios/status.go [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+	q.SetNonEmpty("formatoptions", s.FormatOptions.String())
+	q.SetNonEmpty("dateformat", s.DateFormat)
+	q.SetNonEmpty("hostname", s.HostName)
+	q.SetNonEmpty("servicedescription", s.ServiceDescription)
+
+	return q
+}
+
+func (s ServiceRequest) Build() Query {
+	return s.build("service")
```
Comment (amwolff) [inline nagios/status.go]: ditto.

File: nagios/status.go [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+
+	q.SetNonEmpty("query", query)
+	q.SetNonEmpty("formatoptions", p.FormatOptions.String())
+	q.SetNonEmpty("dateformat", p.DateFormat)
+
+	return q
+}
+
+func (p PerformanceDataRequest) Build() Query {
+	return p.build("performancedata")
```
Comment (amwolff) [inline nagios/status.go]: ditto.

File: nagios/status.go [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+}
+
+type Performance struct {
```
Comment (amwolff) [inline nagios/status.go]: Follow the convention set earlier. Move this type after the `PerformanceData` type.

BTW It's super easy to get lost here. Maybe we could utilize the `type ( ... )` clause to group types used in the same context?

File: nagios/status.go [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+	FormatVersion int             `json:"format_version"`
+	Result        Result          `json:"result"`
+	Data          PerformanceData `json:"data"`
+}
+
+type Checks struct {
+	OneMin  int `json:"1min"`
+	FiveMin int `json:"5min"`
+	One5Min int `json:"15min"`
```
Comment (amwolff) [inline nagios/status.go]: It's fifteen.

File: nagios/status.go [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+	Passive Passive `json:"passive"`
+}
+
+type ActiveScheduledHostChecks struct {
+	OneMin  int `json:"1min"`
+	FiveMin int `json:"5min"`
+	One5Min int `json:"15min"`
+}
+
+type ActiveOndemandHostChecks struct {
```
Comment (amwolff) [inline nagios/status.go]: Use camelCase.

File: nagios/status.go [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+	One5Min int `json:"15min"`
+}
+
+type ActiveScheduledServiceChecks struct {
+	OneMin  int `json:"1min"`
+	FiveMin int `json:"5min"`
+	One5Min int `json:"15min"`
+}
+
+type ActiveOndemandServiceChecks struct {
```
Comment (amwolff) [inline nagios/status.go]: ditto.

File: nagios/status.go [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+	InUse          int `json:"in_use"`
+	MaxUsed        int `json:"max_used"`
+	TotalAvailable int `json:"total_available"`
+}
+
+type BufferUsage struct {
+	ExternalCommands ExternalCommandsBuffer `json:"external_commands"`
+}
+
+type Programstatus struct {
```
Comment (amwolff) [inline nagios/status.go]: ditto.

**Review by amwolff — CHANGES_REQUESTED** [review]

File: nagios/status.go (hunk 4 above)
Comment (DanielSz50) [reply nagios/status.go]: I thought it looks slightly better when we code it this way even if it's unnecessary.

File: nagios/status.go (hunk 7 above)
Comment (DanielSz50) [reply nagios/status.go]: Cool, I didn't know about that functionality.

File: nagios/status.go (hunk 2 above)
Comment (DanielSz50) [reply nagios/status.go:135]: Sure thing, my mistake.

File: nagios/status.go (hunk 1 above)
Comment (DanielSz50) [reply nagios/status.go]: Right.

File: nagios/status.go (hunk 9 above)
Comment (DanielSz50) [reply nagios/status.go]: My mistake.

File: nagios/status.go (hunk 3 above)
Comment (DanielSz50) [reply nagios/status.go]: Right.

**Review by amwolff — DISMISSED** [review]
LGTM, but we mustn't forget tests. We need both unit + `internal/integration`.

The bottom line is: merge this if you need this now, add tests later or add tests now.

**Comment by DanielSz50** [comment]
I will include tests later that day.

File: nagios/status.go [hunk 12]
```diff
@@ -73,20 +73,27 @@ func (g GeneralHostRequest) build(query string, includeStartCount bool) Query {
...
 	q.SetNonEmpty("contactname", g.ContactName)
 
 	q.SetNonEmpty("hosttimefield", g.HostTimeField)
-	q.SetNonEmpty("starttime", strconv.FormatInt(g.StartTime, 10))
-	q.SetNonEmpty("endtime", strconv.FormatInt(g.EndTime, 10))
+
+	if g.StartTime != 0 || g.EndTime != 0 {
+		q.URLQuery.Set("starttime", strconv.FormatInt(g.StartTime, 10))
+		q.URLQuery.Set("endtime", strconv.FormatInt(g.EndTime, 10))
+	}
```
Comment (amwolff) [inline nagios/status.go]: Maybe we should adjust the `SetNonEmpty` method. It shouldn't hurt the rest of the client code, I think.

Of course not requesting that in this patch set.

File: nagios/status.go [hunk 13]
```diff
@@ -125,9 +132,10 @@ type HostCount struct {
 	Data          HostCountData `json:"data"`
 }
 
+//TODO(DanielSz50): Handle showDetails check properly
 type HostListData struct {
 	Selectors map[string]json.RawMessage `json:"selectors"`
-	HostList  map[string]string          `json:"hostlist"`
+	HostList  map[string]json.RawMessage `json:"hostlist"`
```
Comment (amwolff) [inline nagios/status.go:135]: Hm, I'm guessing the `map[string]string` failed to work with `Enumerate` switched. Strange.

File: nagios/status.go [hunk 14]
```diff
@@ -222,15 +234,18 @@ func (g GeneralServiceRequest) build(query string, includeStartCount bool) Query
...
 	q.SetNonEmpty("contactname", g.ContactName)
 
 	q.SetNonEmpty("servicetimefield", g.ServiceTimeField)
-	q.SetNonEmpty("starttime", strconv.FormatInt(g.StartTime, 10))
-	q.SetNonEmpty("endtime", strconv.FormatInt(g.EndTime, 10))
+
+	if g.StartTime != 0 || g.EndTime != 0 {
+		q.URLQuery.Set("starttime", strconv.FormatInt(g.StartTime, 10))
+		q.URLQuery.Set("endtime", strconv.FormatInt(g.EndTime, 10))
+	}
```
Comment (amwolff) [inline nagios/status.go]: I'm not sure whether we should include both times when only one is present.

File: nagios/status.go [hunk 15]
```diff
@@ -294,54 +309,56 @@ func (h HostRequest) Build() Query {
...
+	Status            string `json:"status"`
+	LastUpdate        int64  `json:"last_update"`
+	HasBeenChecked    bool   `json:"has_been_checked"`
+	ShouldBeScheduled bool   `json:"should_be_scheduled"`
+	CurrentAttempt    int    `json:"current_attempt"`
+	MaxAttempts       int    `json:"max_attempts"`
+	LastCheck         int64  `json:"last_check"`
+	NextCheck         int64  `json:"next_check"`
+
+	// CheckOptions with bitmask turns to the array
```
Comment (amwolff) [inline nagios/status.go]:
```suggestion
	// CheckOptions with bitmask turns into an array.
```

File: nagios/status_test.go [hunk 16]
```diff
@@ -0,0 +1,542 @@
...
+				GeneralServiceRequest: tt.GeneralServiceRequest,
+			}
+			if got := a.Build(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Build() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestServieListRequest_Build(t *testing.T) {
```
Comment (amwolff) [inline nagios/status_test.go]:
```suggestion
func TestServiceListRequest_Build(t *testing.T) {
```

BTW you can generate these test if you're using some kind of IDE (although cleaning them up is a bit of work).

File: nagios/status_test.go [hunk 17]
```diff
@@ -0,0 +1,542 @@
...
+				t.Errorf("Build() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestHostRequest_build(t *testing.T) {
+	type args struct {
+		query string
+	}
```
Comment (amwolff) [inline nagios/status_test.go]: It's unused.

File: nagios/status_test.go [hunk 18]
```diff
@@ -0,0 +1,542 @@
...
+				t.Errorf("build() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestServiceRequest_build(t *testing.T) {
+	type args struct {
+		query string
+	}
```
Comment (amwolff) [inline nagios/status_test.go]: ditto.

File: nagios/status_test.go [hunk 19]
```diff
@@ -0,0 +1,542 @@
...
+				t.Errorf("build() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestPerformanceDataRequest_Build(t *testing.T) {
+	type args struct {
+		query string
+	}
```
Comment (amwolff) [inline nagios/status_test.go]: ditto.

**Review by amwolff — CHANGES_REQUESTED** [review]

File: nagios/status.go (hunk 13 above)
Comment (DanielSz50) [reply nagios/status.go:135]: After checking the showDetails option JSON looks completely different.

File: nagios/status.go (hunk 12 above)
Comment (DanielSz50) [reply nagios/status.go]: Adjusted, let me know if it's ok.

File: nagios/query.go [hunk 20]
```diff
@@ -8,6 +8,13 @@ type Query struct {
...
 
 func (q Query) SetNonEmpty(key, value string) {
+	if key == startTimeKey || key == endTimeKey {
+		if value != "0" {
+			q.URLQuery.Set(key, value)
+		}
+		return
+	}
+
 	if len(value) > 0 {
```
Comment (amwolff) [inline nagios/query.go]: I was thinking about something more like this. Exclude all zero-values:
```suggestion
	if len(value) > 0 && value != "0" {
```

File: nagios/status.go [hunk 21]
```diff
@@ -0,0 +1,631 @@
...
+	Count     HostStatusCount            `json:"count"`
+}
+
+type HostCount struct {
+	FormatVersion int           `json:"format_version"`
+	Result        Result        `json:"result"`
+	Data          HostCountData `json:"data"`
+}
+
+//TODO(DanielSz50): Handle showDetails check properly
```
Comment (amwolff) [inline nagios/status.go]:
```suggestion
// TODO(DanielSz50): Handle showDetails check properly.
```

Oh, yeah, we should have that linter enabled.

File: nagios/status.go (hunk 13 above)
Comment (amwolff) [reply nagios/status.go:135]: Ack.

File: nagios/status.go [hunk 22]
```diff
@@ -0,0 +1,631 @@
...
+	PerfData          string `json:"perf_data"`
+	MaxAttempts       int    `json:"max_attempts"`
+	CurrentAttempt    int    `json:"current_attempt"`
+	Status            string `json:"status"`
+	LastUpdate        int64  `json:"last_update"`
+	HasBeenChecked    bool   `json:"has_been_checked"`
+	ShouldBeScheduled bool   `json:"should_be_scheduled"`
+	LastCheck         int64  `json:"last_check"`
+
+	// CheckOptions with bitmask turns to the array
```
Comment (amwolff) [inline nagios/status.go:405]: Forgot about this comment.

**Review by amwolff — CHANGES_REQUESTED** [review]
I was thinking of a different modification to the `SetNonEmpty`. Let me know what you think.

**Comment by DanielSz50** [comment]
I didn't want to accidentally omit numeric values that could be set to 0, but now I see it shouldn't be a problem.

File: nagios/status.go (hunk 12 above)
Comment (amwolff) [reply nagios/status.go]: Looks good to me :+1:

**Review by amwolff — APPROVED** [review]

**Pull request merged**